src/queue/priority.rs

```diff
@@ -0,0 +1,492 @@
+        T: TaskCell,
+    {
+        self.level_manager.adjust_task_level(task_cell);
+        self.priority_manager.get_priority(task_cell.mut_extras())
+    }
+}
+
+/// The global priority queue. We use a `SkipMap` as a priority queue,
+/// The key is the priority and value is the task.
+struct QueueCore<T> {
+    pq: SkipMap<MapKey, Slot<T>>,
+    /// a global sequence generator to ensure all task keys are unique.
+    sequence: AtomicU64,
+}
+
+impl<T> QueueCore<T> {
+    fn new() -> Self {
+        Self {
+            pq: SkipMap::new(),
+            sequence: AtomicU64::new(0),
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.pq.is_empty()
+    }
+}
+
+impl<T: TaskCell + Send + 'static> QueueCore<T> {
+    fn push(&self, msg: T, priority: u64) {
+        self.pq.insert(self.gen_key(priority), Slot::new(msg));
+    }
+
+    pub fn pop(&self) -> Option<Pop<T>> {
+        fn into_pop<T>(mut t: T) -> Pop<T>
+        where
+            T: TaskCell,
+        {
+            let schedule_time = t.mut_extras().schedule_time.unwrap();
+            Pop {
+                task_cell: t,
+                schedule_time,
+                from_local: false,
+            }
+        }
+
+        self.pq
+            .pop_front()
+            .map(|e| into_pop(e.value().take().unwrap()))
+    }
+
+    #[inline]
+    fn gen_key(&self, weight: u64) -> MapKey {
+        MapKey(weight, self.sequence.fetch_add(1, Ordering::Relaxed))
+    }
+}
+
+/// A holder to store task. Wrap the task in a AtomicCell becuase crossbeam-skip only provide
+/// readonly acess to a popped Entry.
+struct Slot<T> {
+    cell: AtomicCell<Option<T>>,
+}
+
+impl<T> Slot<T> {
+    fn new(value: T) -> Self {
+        Self {
+            cell: AtomicCell::new(Some(value)),
+        }
+    }
+
+    fn take(&self) -> Option<T> {
+        self.cell.take()
+    }
+}
+
+/// The runner for priority queues.
+///
+/// The runner helps collect additional information to support auto-adjust task level.
+/// [`PriorityRunnerBuiler`] is the [`RunnerBuilder`] for this runner.
+pub struct PriorityRunner<R> {
+    inner: R,
+    local_level0_elapsed_us: LocalIntCounter,
+    local_total_elapsed_us: LocalIntCounter,
+}
+
+impl<R, T> Runner for PriorityRunner<R>
+where
+    R: Runner<TaskCell = T>,
+    T: TaskCell,
+{
+    type TaskCell = T;
+
+    fn start(&mut self, local: &mut Local<T>) {
+        self.inner.start(local)
+    }
+
+    fn handle(&mut self, local: &mut Local<T>, mut task_cell: T) -> bool {
+        let extras = task_cell.mut_extras();
+        let level = extras.current_level();
+        let total_running_time = extras.total_running_time.clone();
+        let begin = Instant::now();
+        let res = self.inner.handle(local, task_cell);
+        let elapsed = begin.elapsed();
+        let elapsed_us = elapsed.as_micros() as u64;
+        if let Some(ref running_time) = total_running_time {
+            running_time.inc_by(elapsed);
+        }
+        if level == 0 {
```

Comment:
shouldn't always be 0?

Reply:
No. The priority runner also implements the level degradation logic like the multilevel queue, so the user may assign a different priority for higher level tasks.